protocol/x/accountplus/ante/ante.go

```diff
@@ -0,0 +1,299 @@
+	err = ad.ValidateAuthenticatorFeePayer(tx)
+	if err != nil {
+		return sdk.Context{}, err
+	}
+
+	msgs := tx.GetMsgs()
+	if len(msgs) == 0 {
+		return ctx, errorsmod.Wrap(sdkerrors.ErrInvalidRequest, "no messages in transaction")
+	}
+
+	feeTx, ok := tx.(sdk.FeeTx)
+	if !ok {
+		return ctx, errorsmod.Wrap(sdkerrors.ErrTxDecode, "Tx must be a FeeTx")
+	}
+
+	// The fee payer is the first signer of the transaction. This should have been enforced by the
+	// LimitFeePayerDecorator
+	signers, _, err := ad.cdc.GetMsgV1Signers(msgs[0])
+	if err != nil {
+		return ctx, errorsmod.Wrap(sdkerrors.ErrUnauthorized, "failed to get signers")
+	}
+	feePayer := sdk.AccAddress(signers[0])
+	feeGranter := feeTx.FeeGranter()
+	fee := feeTx.GetFee()
+
+	selectedAuthenticators, err := ad.GetSelectedAuthenticators(tx, len(msgs))
+	if err != nil {
+		return ctx, err
+	}
+
+	// tracks are used to make sure that we only write to the store after every message is successful
+	var tracks []func() error
+
+	// Authenticate the accounts of all messages
+	for msgIndex, msg := range msgs {
+		signers, _, err := ad.cdc.GetMsgV1Signers(msg)
+		if err != nil {
+			return ctx, errorsmod.Wrap(sdkerrors.ErrUnauthorized, "failed to get signers")
+		}
+		// Enforce only one signer per message
+		if len(signers) != 1 {
+			return sdk.Context{}, errorsmod.Wrap(sdkerrors.ErrUnauthorized, "messages must have exactly one signer")
+		}
+
+		// By default, the first signer is the account that is used
+		account := sdk.AccAddress(signers[0])
+
+		// Get the currently selected authenticator
+		selectedAuthenticatorId := selectedAuthenticators[msgIndex]
+		selectedAuthenticator, err := ad.accountPlusKeeper.GetInitializedAuthenticatorForAccount(
+			ctx,
+			account,
+			selectedAuthenticatorId,
+		)
+		if err != nil {
+			return sdk.Context{},
+				errorsmod.Wrapf(
+					err,
+					"failed to get initialized authenticator "+
+						"(account = %s, authenticator id = %d, msg index = %d, msg type url = %s)",
+					account,
+					selectedAuthenticatorId,
+					msgIndex,
+					sdk.MsgTypeURL(msg),
+				)
+		}
+
+		// Generate the authentication request data
+		authenticationRequest, err := authenticator.GenerateAuthenticationRequest(
+			ctx,
+			ad.cdc,
+			ad.accountKeeper,
+			ad.sigModeHandler,
+			account,
+			feePayer,
+			feeGranter,
+			fee,
+			msg,
+			tx,
+			msgIndex,
+			simulate,
+		)
+		if err != nil {
+			return sdk.Context{},
+				errorsmod.Wrapf(
+					err,
+					"failed to generate authentication data "+
+						"(account = %s, authenticator id = %d, msg index = %d, msg type url = %s)",
+					account,
+					selectedAuthenticator.Id,
+					msgIndex,
+					sdk.MsgTypeURL(msg),
+				)
+		}
+
+		a11r := selectedAuthenticator.Authenticator
+		stringId := strconv.FormatUint(selectedAuthenticator.Id, 10)
+		authenticationRequest.AuthenticatorId = stringId
+
+		// Consume the authenticator's static gas
+		ctx.GasMeter().ConsumeGas(a11r.StaticGas(), "authenticator static gas")
```

Comment:
To confirm, this is just the static portion but not the full gas cost right (inferred from this [comment](https://github.com/dydxprotocol/v4-chain/blob/c34d56d6e395e8311556883c3b3afe67f25d47f7/protocol/x/accountplus/authenticator/signature_authenticator.go#L37))

Reply:
yeah - since our validations are mostly stateless, gas consumption (static + gas incurred during `Authenticate`) are going to be 0 for most of the validations. `SignatureVerification` incurs `SigVerifyCostSecp256k1` though

Reply:
thinking about this more, seems like we are double charging gas [here](https://github.com/dydxprotocol/v4-chain/blob/main/protocol/x/accountplus/authenticator/signature_authenticator.go#L60) and [here](https://github.com/dydxprotocol/v4-chain/blob/main/protocol/app/ante.go#L137).

removed gas for the new `SignatureAuthenticator`. let me know if you have objections.

Reply:
It looks like Osmosis way of solving this is to create two separate chain of decorators, and only include `SigGasConsumeDecorator` in the [classic chain](https://github.com/osmosis-labs/osmosis/blob/4ce0dbf17e0fffbecdcf5781b50cb9334acd4b1b/app/ante.go#L73). Did we consider this approach?

Also, is there any other modifications needed for the new route? For instance Osmosis code implies this [decorator](https://github.com/osmosis-labs/osmosis/blob/4ce0dbf17e0fffbecdcf5781b50cb9334acd4b1b/app/ante.go#L71C8-L71C29) is only needed for the classic route.

Reply:
I was thinking for simplicity we can just have 
```
sigVerification: accountplusante.NewCircuitBreakerDecorator(
			options.AccountplusKeeper,
			accountplusante.NewAuthenticatorDecorator( // this is the new ante handler
				...
			),
			customante.NewSigVerificationDecorator( // this is the old sigverify (without replay protection)
				...
			),
		),
```
i.e. without the two chained decorators and just share everything else except for sigverify. Agree that things like `NewSetPubKeyDecorator` is only relevant for old approach, but I feel like that's an optimization that we can worry about later. wdyt?

Reply:
as for the gas behavior tests, created [this](https://linear.app/dydx/issue/CT-1238/write-tests-to-verify-gas-behavior-of-the-new-decorator) ticket to track